### ISSUE TITLE
[Feature Request] Allow HTTP filters to add a body that's not readily available to headers-only responses 

### ISSUE THREAD

**Issue by yosrym93** [envoyproxy/envoy#12814]
**TL;DR:** For headers-only responses, we need a way to tell the FilterManager in `encodeHeaders` that we want to add body to this response without having it readily available (i.e don't end stream after sending headers downstream).

I am working on the HTTP CacheFilter. When a cached response is being revalidated, a 304 (not modified) response (headers only) means that the cached response is valid. In this scenario, the CacheFilter (currently) updates the headers in `encodeHeaders`,  stops iteration, and requests the cached response from the cache. The cache fetches the cached response along several chunks of data. Whenever a chunk is ready, the CacheFilter calls `StreamEncoderFilterCallbacks::addEncodedData` to add it to the body. After the whole cached response is added, the CacheFilter calls `StreamEncoderFilterCallbacks::continueEncoding` to resume encoding with the full response.

This approach has several problems:
- The headers and body are held until the response fully available, which incurs a delay. 
- The memory usage for buffering the whole response may be high, and it may cause a buffer overflow.

Ideally we would stream headers and body chunks downstream whenever they are available. We could use `injectEncodedDataToFilterChain` instead of `addEncodedData` to directly pass the data chunks as they arrive. However, the problem is (AFAIK) there's no current way in the filters api to "tell" the HCM (or the `FilterManager`) in `encodeHeaders` that there's body that will be added to this response without having it ready. In other words, change `end_stream` to false.

One way of changing a headers-only response to a response w/ body without making much changes to the HCM or the API is to call `addEncodedData` in `CacheFilter::encodeHeaders` with an empty body. This will cause data iteration for the response to kick-start after the headers iteration. However, the way `addEncodedData` works now is that if a filter calls it during the `encodeHeaders` callback, and it continues iteration, then the data iteration starts at the next filter downstream. This means the CacheFilter has no way of stopping the data iteration until the cached response is fetched.

If acceptable, the default behavior of `addEncodedData` in this case (being called in `encodeHeaders`) can be changed by a small edit to `FilterManager::encodeHeaders`. Specifically changing this if condition:
```
// Here we handle the case where we have a header only response, but a filter adds a body
// to it. We need to not raise end_stream = true to further filters during inline iteration.
if (end_stream && buffered_response_data_ && continue_data_entry == encoder_filters_.end()) {
    continue_data_entry = entry;
}
``` 
to:
```
// Here we handle the case where we have a header only response, but a filter adds a body
// to it. We need to not raise end_stream = true to further filters during inline iteration.
if (end_stream && buffered_response_data_ && continue_data_entry == encoder_filters_.end()) {
    continue_data_entry = entry;
    (*entry)->end_stream_ = false;
    (*entry)->iterate_from_current_filter_ = true;
}
```

Currently, only a single HTTP filter (json_transcoder) calls `addEncodedData` during `encodeHeaders`. If this default behavior is acceptable we can implement it and fix this filter (maybe by skipping `encodeData` if it was invoked by the filter itself).

Another way to do this is to change the `addEncodedData` api and add an extra parameter as whether the filter wants to participate in encoding the data it added. However, (AFAIK) this parameter would only be relevant when the function is called in `encodeHeaders` callbacks, which is not the common case.

Any other suggestions on ways to accomplish this are highly appreciated!

**Comment by yosrym93** [comment]
@alyssawilk

**Comment by yosrym93** [comment]
@capoferro @toddmgreer

**Comment by alyssawilk** [comment]
I like the idea of using addEncodedData, but I wonder if rather than changing it to call encodeData for the current filter it'd make more sense to add an end_stream boolean argument (which could default true, or be manually set true for the one use case) to allow for multiple addEncodedData chunks until the cache response is complete.  I'm concerned with the proposed change, there's still no way for a filter to indicate when the added encoded data is complete, so one might need to pause processing until the entire body were local, where it'd be better to stream the response for large cached responses.  
cc @mattklein123 for thoughts

**Comment by yosrym93** [comment]
The problem with using `addEncodedData` is that it only appends data to the buffer. It does not run it through filters. Also, my understanding is that there's more complexity to how `end_stream` is handled (there is local `end_stream`s for each filter's ActiveStream wrappers). That's why I think using `injectEncodedDataToFilterChain` is the right choice here. However, we would still need a way to make the stream not end after the headers are encoded.

The proposed change is only intended to make the headers streaming go all the way downstream before the response processing is stopped waiting for the cached body. If there's another way to "tell" the FilterManager to "continue encoding headers but then wait because data is coming" then it should work too.

**Comment by yosrym93** [comment]
I suppose if an `end_stream` parameter is added to `addEncodedData` we can use it instead of `injectEncodedDataToFilterChain` by calling `addEncodedData(data, end_stream)` then `continueEncoding` each time a chunk is available. But I don't know how that `end_stream` parameter would work.

**Comment by toddmgreer** [comment]
Side note: We'll need to handle the (noncompliant, but possible) case of the upstream 304 response including a body. We could either discard the body or turn it into some kind of error. It's possible that Envoy's framing code already does this.

**Comment by yosrym93** [comment]
> Side note: We'll need to handle the (noncompliant, but possible) case of the upstream 304 response including a body. We could either discard the body or turn it into some kind of error. It's possible that Envoy's framing code already does this.

In the proposed approach we should return `StopIterationNoBuffer` from `encodeData` anyway, which I assume will drop any buffered body.

**Comment by alyssawilk** [comment]
both run through the filter chain, but to your point if you don't want HCM buffering (and you have to do some internal buffering either way) injectEncodedDataToFilterChain is fine.  I'm less concerned about which function you use,  and more concerned about the plan to start iterating from the current filter.  Maybe I'm just not understanding why you want to halt the flow of data?    I'd think if you were streaming a large number of chunks you'd want the initial chunks being sent downstream as soon as they're available, to avoid an OOM vector for large cache responses.

**Comment by yosrym93** [comment]
>  Maybe I'm just not understanding why you want to halt the flow of data? I'd think if you were streaming a large number of chunks you'd want the initial chunks being sent downstream as soon as they're available, to avoid an OOM vector for large cache responses.

I do not want to half the stream of data. I only want to half the empty body that would be added by `encodeHeaders`. Then whenever a chunk of cached data is ready we can stream it downstream using `injectEncodedDataToFilterChain`.

**Comment by yosrym93** [comment]
I only want to start iterating from the current filter to invoke `CacheFilter::encodeData` with the empty body to be able to halt it and stop the response encoding. Then start encoding cached chunks of data as they arrive.

**Comment by alyssawilk** [comment]
I would think that injectEncodedData with no data but end_stream false would do what you want - it should inform the HCM that there's now a body with the request, and that the body is incomplete.  I wouldn't assume it works without testing it, but I think that would be the ideal way to handle this.

**Comment by yosrym93** [comment]
`injectEncodedData` directly calls `FilterManager::encodeData` which calls subsequent filters' `encodeData`. According to its documentation it should not be used in callback context. Also, subsequent filters then would have `encodeData` called before `encodeHeaders` which might cause problems.

**Comment by toddmgreer** [comment]
When we have a 10GiB response in a remote cache, we have to request it in small chunks and send each chunk downstream as soon as we get it. We will call injectEncodedData (or whatever) many times, as soon as the data comes in from the cache. When downstream gets backed up, we need to wait until it's clear before requesting the next chunk from the cache.

**Comment by yosrym93** [comment]
@alyssawilk The main purpose of this issue is "how to tell the HCM in `encodeHeaders` that more data is coming without having it at the moment (i.e don't end stream), then let the headers flow downstream".

Modifying the behavior of calling `addEncodedData` in `encodeHeaders` to use it with an empty body in `CacheFilter::encodeHeaders` and then stopping iteration in `CacheFilter::encodeData` is only a suggested way to accomplish this. As this will cause headers to flow downstream and the data encoding will be stopped until chunks of cached body arrive.

If there are any better or cleaner ways to do this, or that require less (or no) modifications to the FilterManager/HCM, that would be great!

**Comment by mattklein123** [comment]
@yosrym93 I believe I already discussed this issue with you in Slack. I believe using `injectEncodedDataToFilterChain` is the right solution here. Can you summarize the issue with that API? I don't think there is any issue using this in callback context as this is what the fault rate limit filter does. It's called in timer context.

So is the issue that if you set end_stream = false in that API, it's broken because you have not sent headers (stopped in headers). and headers are not sent when you inject with end_stream = false? I think we can fix that.

**Comment by yosrym93** [comment]
@mattklein123  In summary, we need a way in `CacheFilter::encodeHeaders` to continue iteration without ending the stream. We want the headers iteration to continue downstream, but then wait until cached chunks start coming in and send them downstream using `injectEncodedDataToFilterChain`.

**Comment by mattklein123** [comment]
Can't you just send the headers with the first cached chunk for now? Yes it's not perfect but it's probably good enough with a TODO. Does that work without changes?

**Comment by yosrym93** [comment]
> Can't you just send the headers with the first cached chunk for now? Yes it's not perfect but it's probably good enough with a TODO. Does that work without changes?

How would we do that? We could stop iteration in `encodeHeaders` until the first chunk arrives. Then we could add the chunk using `addEncodedData` and `continueEncoding`. If we do that, we will face the same problem, we want to till the HCM this is not the end of the stream to wait for the next chunks.

AFAIK, If we instead use `injectEncodedDataToFilterChain` with the first chunk, then that chunk will be sent downstream before the headers.

**Comment by mattklein123** [comment]
> AFAIK, If we instead use injectEncodedDataToFilterChain with the first chunk, then that chunk will be sent downstream before the headers.

Right I would argue this is a bug. Can we just fix this?

**Comment by yosrym93** [comment]
> Right I would argue this is a bug. Can we just fix this?

Is the behavior you are suggesting as the correct behavior is:
Before calling `FilterManager::encodeData`, if headers iteration was stopped, continue headers iteration first?

If so, what happens if another filter downstream stops headers iteration again? The data passed to `injectEncodedDataToFilterChain` is not added to the buffer so what happens then?

Also, what happens if the headers iteration was completed, but there was stopped data iteration on already buffered data. Should we continue that or drop it or ignore it?

I think modifying `injectEncodedDataToFilterChain` behavior will make it more complex and difficult to maintain.

**Comment by yosrym93** [comment]
I don't believe this is a bug, it's the way the function is implemented. It simply takes a body and directly passes it through the filter chain.

**Comment by mattklein123** [comment]
> I don't believe this is a bug, it's the way the function is implemented. It simply takes a body and directly passes it through the filter chain.

I implemented it. ;)

I think it's a bug. The way it should work IMO is that if you inject data and headers have not been continued, headers are continued, and then data is injected. If a further filter stops that really doesn't matter. The HCM will take care of it.

The other option would be to add a new response code to `FilterHeadersStatus`, something like `ContinueAndDontEndStream`, to match `ContinueAndEndStream`. This would also do what you want and is inline with the other APIs.

**Comment by yosrym93** [comment]
> I implemented it. ;)

Well then sorry for any wrong assumption I made about it's intention!
 
> I think it's a bug. The way it should work IMO is that if you inject data and headers have not been continued, headers are continued, and then data is injected. If a further filter stops that really doesn't matter. The HCM will take care of it.

Can you clarify what would happen here? My understanding is that you mean something like:
```
if (headers not completed) {
   doHeaders(false);
}
parent_.encodeData(this, data, end_stream, FilterManager::FilterIterationStartState::CanStartFromCurrent);
```
What would happen then if another filter stopped headers iteration? Would the HCM check it's iteration status before calling its `encodeData` callback? If so, what does the HCM do since this body is not buffered anywhere? Does it buffer it?

> The other option would be to add a new response code to `FilterHeadersStatus`, something like `ContinueAndDontEndStream`, to match `ContinueAndEndStream`. This would also do what you want and is inline with the other APIs.

I think this could be a cleaner solution. I will look further into it!

**Comment by toddmgreer** [comment]
If we use injectEncodedDataToFilterChain, I presume we'll need to modify it to return FilterDataStatus::StopIterationAndWatermark (or similar) when the downstream buffers are full. Otherwise CacheFilter will just keep pulling more chunks from cache until we run out of either response body or memory.

**Comment by yosrym93** [comment]
> If we use injectEncodedDataToFilterChain, I presume we'll need to modify it to return FilterDataStatus::StopIterationAndWatermark (or similar) when the downstream buffers are full. Otherwise CacheFilter will just keep pulling more chunks from cache until we run out of either response body or memory.

According my limited understanding of: https://github.com/envoyproxy/envoy/blob/master/source/docs/flow_control.md#encoder-filters
I think to handle downstream backpressure we need to register to receive watermark events, and implement `DownstreamWatermarkCallbacks::onAboveWriteBufferHighWatermark()` and `DownstreamWatermarkCallbacks::onBelowWriteBufferLowWatermark()`. Which I assume will stop and resume fetching body chunks from cache.

**Comment by mattklein123** [comment]
> Can you clarify what would happen here? My understanding is that you mean something like:

Yes

> What would happen then if another filter stopped headers iteration? Would the HCM check it's iteration status before calling its encodeData callback? If so, what does the HCM do since this body is not buffered anywhere? Does it buffer it?

Buffering or not buffering is on a per filter basis. If one filter injects, and another buffers, that's fine. That should be handled by the HCM without much work, if any.

> I think to handle downstream backpressure we need to register to receive watermark events,

Yes, no matter what you do you will need to register for watermark callbacks and apply back pressure on the cache fetch.

**Comment by yosrym93** [comment]
> Buffering or not buffering is on a per filter basis. If one filter injects, and another buffers, that's fine. That should be handled by the HCM without much work, if any.

I am not sure I understand this. My question is, what happens if we change `injectEncodedDataToFilterChain` to doHeaders first if the headers are not yet completed (stopped by some filter). Then another filters stops headers iteration again. Wouldn't this cause the same problem? Some filters will receive data before headers?

Also, isn't it incorrect to force continue headers iteration after some other filter stopped iteration?

**Comment by mattklein123** [comment]
Each filters iteration state is on a per filter basis. What one filter does has no basis on further filters. If you continue headers, and some other filter stops it, that is tracked by the HCM. You don't need to deal with this.

**Comment by yosrym93** [comment]
Then why it is a bug when a filter stops headers iteration then another filter injects data?

**Comment by mattklein123** [comment]
> Then why it is a bug when a filter stops headers iteration then another filter injects data?

What we are discussing has nothing to do with multiple filters. It has to do with a single filter. We are solving the problem of how to deal with injecting data when headers have not been continued. You can either do this implicitly as I have suggested, or do it explicitly with a new return code from the headers callback.

**Comment by yosrym93** [comment]
Oh yeah I got it. So basically if a filter stops headers iteration then tries to inject data we should implicitly continue the headers iteration. Yeah that sounds reasonable.
Anyway I think I will work on implementing the second solution, as it would allow the headers to flow through without having to wait for the second chunk. I will add a TODO for the though so that anyone can get back to it later.
Thanks for your help!

**Comment by yosrym93** [comment]
What types of tests do I need to write for the new `FilterHeaderStatus::ContinueDontEndStream` ? 
I looked around the code and I think I can add a test to "https://github.com/envoyproxy/envoy/tree/master/test/integration/filters" of a filter that receives a headers-only request/response, continues headers iteration without ending the stream, then after an alarm or something either:
- Adds encoded/decoded data and continues iteration, or
-  Injects encoded/decoded data
to add a body to the headers-only request/response.  
Does this sound right? Do I need to write other tests?

**Comment by mattklein123** [comment]
> Does this sound right? Do I need to write other tests?

You should add both unit tests in conn_manager_impl_test as well as integration tests as you outline. Thank you!
